### PR TITLE
travis: create config

### PR DESCRIPTION
**Pull request by xoviat**
Unfortunately I was not able to get osx working. There is a fairly significant compiler error that involves undefined symbols, and I don't understand osx well enough to resolve the problem.

### CODE REVIEW

**Comment by xoviat** [comment]
And also I could not install 32 bit Python at all on osx.

File: scripts/prepare-linux.sh [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ln -s /usr/bin/ccache $CCACHE_DIR/c++
+export PATH=$CCACHE_DIR:$PATH
+which gcc
+
+# Install a newer CMake
+if [ $TARGET_ARCH == "x64" ]; then
+  CMAKE_URL="https://cmake.org/files/v3.5/cmake-3.5.2-Linux-x86_64.tar.gz"  
+else
+  CMAKE_URL="https://cmake.org/files/v3.5/cmake-3.5.2-Linux-i386.tar.gz"  
+fi
```
Comment (jcfr) [inline scripts/prepare-linux.sh]: you could also pip install cmake now.

File: scripts/prepare-linux.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+DEPS_DIR="$PWD/deps"
+
+# Use and verify ccache
+CCACHE_DIR=$DEPS_DIR/ccache
+mkdir -p $CCACHE_DIR
+ln -s /usr/bin/ccache $CCACHE_DIR/gcc
+ln -s /usr/bin/ccache $CCACHE_DIR/g++
+ln -s /usr/bin/ccache $CCACHE_DIR/cc
+ln -s /usr/bin/ccache $CCACHE_DIR/c++
+export PATH=$CCACHE_DIR:$PATH
```
Comment (jcfr) [inline scripts/prepare-linux.sh]: To streamline the setup, you could use dockcross-manylinux images.
See https://github.com/dockcross/dockcross

We could easily include a script named `setup-ccache.sh` that would take care of creating the symlinks. Also ninja and cmake are already available in these images.

File: scripts/prepare-osx.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
+set -e -x
+
+# Due to arrogance on the part of @yyuu, we need to use some bash here (pyenv/pyenv#602)
+ALLOWED_VERSIONS="$TRAVIS_PYTHON_VERSION.[0-9]"
+CANDIDATE_VERSIONS="$(pyenv install -l | grep -Eio $ALLOWED_VERSIONS)"
+SELECTED_VERSION="$(echo ${CANDIDATE_VERSIONS} | tr " " "\n" | grep -v - | tail -1)"
+echo $SELECTED_VERSION
+
+if [ $TARGET_ARCH == "x86" ]; then
+  env PYTHON_CONFIGURE_OPTS="--enable-shared" CFLAGS="-m32" LDFLAGS="-m32" pyenv install $SELECTED_VERSION
```
Comment (jcfr) [inline scripts/prepare-osx.sh]: since 32-bit osx is quite old, I wouldn't bother.

File: scripts/prepare-linux.sh (hunk 2 above)
Comment (xoviat) [reply scripts/prepare-linux.sh]: Honestly setting up linux image takes very little time.

File: scripts/prepare-linux.sh (hunk 1 above)
Comment (xoviat) [reply scripts/prepare-linux.sh]: Nice!

Comment (xoviat) [reply scripts/prepare-linux.sh]: I will try installing using pip, but one issue that I encountered was that the ABI for cmake 3.9 was too new.

**Comment by xoviat** [comment]
The dockcross images are 480MB, while the quay.io images are 300MB. I'm skeptical that there would be any significant speedup from using dockcross.

Do you know what could be happening on osx? It looks like it needs to link against libpython, but cmake isn't passing it. Although my familiarity with osx is minimal at best (mostly travis).

```
FAILED: lib/libvtkWrapping-8.1.1.dylib 
: && /usr/local/opt/ccache/libexec/c++  -O3 -DNDEBUG  -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 1.0.0 -current_version 1.0.0 -o lib/libvtkWrapping-8.1.1.dylib -install_name /Users/travis/build/jcfr/VTKPythonPackage/_skbuild/cmake-build/VTK-build/lib/libvtkWrapping-8.1.1.dylib Filters/Python/CMakeFiles/vtkFiltersPythonObjects.dir/vtkPythonAlgorithm.cxx.o CMakeFiles/vtkWrapping.dir/vtkWrappingKit.cxx.o  lib/libvtkWrappingPython27Core-8.1.1.dylib lib/libvtkCommon-8.1.1.dylib lib/libvtksys-8.1.1.dylib && :
Undefined symbols for architecture x86_64:
  "_PyCallable_Check", referenced from:
      vtkPythonAlgorithm::SetPythonObject(_object*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::ProcessRequest(vtkInformation*, vtkInformationVector**, vtkInformationVector*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillInputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillOutputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
  "_PyErr_Clear", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PyErr_Fetch", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PyImport_ImportModule", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PyInt_AsLong", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PyInt_FromLong", referenced from:
      vtkPythonAlgorithm::FillInputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillOutputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
  "_PyObject_Call", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::SetPythonObject(_object*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::ProcessRequest(vtkInformation*, vtkInformationVector**, vtkInformationVector*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillInputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillOutputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
  "_PyObject_GetAttrString", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::SetPythonObject(_object*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::ProcessRequest(vtkInformation*, vtkInformationVector**, vtkInformationVector*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillInputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillOutputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
  "_PyObject_Str", referenced from:
      vtkPythonAlgorithm::PrintSelf(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, vtkIndent) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PySequence_Fast", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PySequence_Size", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PyString_AsString", referenced from:
      vtkPythonAlgorithm::PrintSelf(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, vtkIndent) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
  "_PyTuple_New", referenced from:
      vtkPythonAlgorithm::CheckResult(char const*, vtkSmartPyObject const&) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::SetPythonObject(_object*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::ProcessRequest(vtkInformation*, vtkInformationVector**, vtkInformationVector*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillInputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
      vtkPythonAlgorithm::FillOutputPortInformation(int, vtkInformation*) in vtkPythonAlgorithm.cxx.o
ld: symbol(s) not found for architecture x86_64
```

**Comment by xoviat** [comment]
After reviewing the docker images, additional dependencies are 50MB at most (the largest is CMake), so using the dockcross image will probably not yield additional speedup.

**Pull request closed without merging**